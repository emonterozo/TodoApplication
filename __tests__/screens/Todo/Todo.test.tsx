import { useRoute } from '@react-navigation/native';

import { act, fireEvent, screen, waitFor } from '../../test-utils';
import { mockSetTodo, renderMockGlobalContextProvider } from '../../context-utils';
import { Todo } from '@app/screens';
import { mockedGoBack } from '../../../jest.setup';


describe('Todo Screen', () => {
  it('should matches the snapshot', () => {
    (useRoute as jest.Mock).mockReturnValue({
      params: {
        todo: undefined,
      },
    });
    renderMockGlobalContextProvider({
      children: <Todo />,
    });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should add new todo', async () => {
    (useRoute as jest.Mock).mockReturnValue({
      params: {
        todo: undefined,
      },
    });
    renderMockGlobalContextProvider({
      children: <Todo />,
    });

    fireEvent.changeText(screen.getByTestId('title-input'), 'Sample Todo');

    fireEvent(screen.getByTestId('add-edit-todo'), 'press');

    await waitFor(() => {
      expect(mockedGoBack).toHaveBeenCalledWith();
    });
  });

  it('should update todo', async () => {
    (useRoute as jest.Mock).mockReturnValue({
      params: {
        todo: { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
      },
    });
    renderMockGlobalContextProvider({
      children: <Todo />,
      todo: [
        { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
      ],
    });

    fireEvent.changeText(screen.getByTestId('title-input'), 'Sample Todo');
    fireEvent.changeText(screen.getByTestId('due-date-input'), '11/12/2024');
    fireEvent(screen.getByTestId('selection-Completed'), 'press');

    fireEvent(screen.getByTestId('add-edit-todo'), 'press');

    await waitFor(() => {
      expect(mockSetTodo).toHaveBeenCalledWith([
        { id: '1234', title: 'Sample Todo', dueDate: '11/12/2024', status: 'Completed' },
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
      ]);
      expect(mockedGoBack).toHaveBeenCalledWith();
    });
  });

  it('should delete todo', async () => {
    (useRoute as jest.Mock).mockReturnValue({
      params: {
        todo: { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
      },
    });
    renderMockGlobalContextProvider({
      children: <Todo />,
      todo: [
        { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
      ],
    });

    fireEvent(screen.getByTestId('delete-todo'), 'press');

    await waitFor(() => {
      expect(mockSetTodo).toHaveBeenCalledWith([
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
      ]);
      expect(mockedGoBack).toHaveBeenCalledWith();
    });
  });
});
