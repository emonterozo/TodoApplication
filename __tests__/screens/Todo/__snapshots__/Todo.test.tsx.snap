// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Screen should matches the snapshot 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
>
  <View
    style={
      {
        "gap": 12,
        "marginVertical": 12,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter description"
      style={
        {
          "borderWidth": 1,
          "height": 40,
          "marginHorizontal": 12,
          "padding": 10,
        }
      }
      testID="title-input"
      value=""
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Due date (MM/DD/YYYY)"
      style={
        {
          "borderWidth": 1,
          "height": 40,
          "marginHorizontal": 12,
          "padding": 10,
        }
      }
      testID="due-date-input"
      value="11/11/2024"
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 5,
              "marginHorizontal": 10,
              "padding": 15,
            },
            {
              "backgroundColor": "green",
            },
          ]
        }
        testID="selection-Pending"
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 18,
            }
          }
        >
          Pending
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 5,
              "marginHorizontal": 10,
              "padding": 15,
            },
            {
              "backgroundColor": "white",
            },
          ]
        }
        testID="selection-In Progress"
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 18,
            }
          }
        >
          In Progress
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 5,
              "marginHorizontal": 10,
              "padding": 15,
            },
            {
              "backgroundColor": "white",
            },
          ]
        }
        testID="selection-Completed"
      >
        <Text
          style={
            {
              "color": "black",
              "fontSize": 18,
            }
          }
        >
          Completed
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 5,
            "marginHorizontal": 10,
            "padding": 15,
          },
          {
            "backgroundColor": "gray",
          },
        ]
      }
      testID="add-edit-todo"
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 20,
          }
        }
      >
        Add
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
